crates/bevy_ecs/src/reflect/bundle.rs

```diff
@@ -274,3 +168,57 @@ impl<C: Bundle + Reflect + FromWorld> FromType<C> for ReflectBundle {
         })
     }
 }
+
+fn insert_field<C: 'static>(entity: &mut EntityMut, field: &dyn Reflect, registry: &TypeRegistry) {
+    if let Some(reflect_component) = registry.get_type_data::<ReflectComponent>(field.type_id()) {
+        reflect_component.apply(entity, field);
+    } else if let Some(reflect_bundle) = registry.get_type_data::<ReflectBundle>(field.type_id()) {
+        reflect_bundle.apply(entity, field, registry);
+    } else {
+        entity.world_scope(|world| {
+            if let Some(id) = world.bundles().get_id(TypeId::of::<C>()) {
+                let info = world.bundles().get(id).unwrap();
+                if info.components().is_empty() {
+                    panic!(
+                        "no `ReflectComponent` registration found for `{}`",
+                        field.type_name()
+                    );
+                }
+            };
+        });
+
+        panic!(
+            "no `ReflectBundle` registration found for `{}`",
+            field.type_name()
+        )
+    }
+}
+
+fn apply_or_insert_field<C: 'static>(
+    entity: &mut EntityMut,
+    field: &dyn Reflect,
+    registry: &TypeRegistry,
+) {
+    if let Some(reflect_component) = registry.get_type_data::<ReflectComponent>(field.type_id()) {
+        reflect_component.apply_or_insert(entity, field);
+    } else if let Some(reflect_bundle) = registry.get_type_data::<ReflectBundle>(field.type_id()) {
+        reflect_bundle.apply_or_insert(entity, field, registry);
+    } else {
+        entity.world_scope(|world| {
+            if let Some(id) = world.bundles().get_id(TypeId::of::<C>()) {
+                let info = world.bundles().get(id).unwrap();
+                if info.components().is_empty() {
+                    panic!(
+                        "no `ReflectComponent` registration found for `{}`",
+                        field.type_name()
+                    );
+                }
+            };
+        });
```

Comment:
What is this for? I feel like this only panics for the empty tuple or bundles with no fields.

Reply:
Looks like it's for a nicer panic message to tell user if he missing `ReflectBundle` or `ReflectComponent`. Am I right, @MrGVSV?

Reply:
The code might assume that if the bundle has no component as per return value of [`components`](https://docs.rs/bevy/latest/bevy/ecs/bundle/struct.BundleInfo.html#method.components), it is therefore a `Component` itself? It's a really roundabout way of doing it. And is it true? I would suspect components to have one component as bundle.

Reply:
Yeah, it's a bit confusing. I just too this part from the #6344, this is why I asking @MrGVSV.

Reply:
Yeah this was meant to provide a nicer panic message. Basically, if we can't find a `ReflectBundle` or `ReflectComponent` for `T` (I had it as `C` but to avoid confusion, I'll refer to it as `T`), then we need to panic.

My goal was to indicate that `T` was missing either `ReflectBundle` or `ReflectComponent` depending on whether or not `T` is a `Bundle` or `Component`. Since `Bundle` is implemented for all `Component` types, I opted to fetch from the `BundleInfo` to check if `T` contained any sub-components. If it didn't, we tell the user they need to register `ReflectComponent` (since a component won't contain sub-components). And if it did, then we tell the user to register `ReflectBundle`.

That being said, I'm not sure why I went with this approach over `World::components`. And I guarantee there are other, better ways of doing this that I just didn't know about haha

Reply:
> That being said, I'm not sure why I went with this approach over World::components

So I can just check if it's a component by using `World::components` and then `Components::get_id` instead?

Reply:
> So I can just check if it's a component by using `World::components` and then `Components::get_id` instead?

I believe so. It should at least be cleaner since we don't need to do an awkward "does this bundle contain components?" check. Again, there may be a better way, but that's just what I can think of.

Reply:
Makes sense to me, done.